packages/common/infra/src/modules/doc/stores/doc-properties.ts

```diff
@@ -0,0 +1,229 @@
+import { differenceBy, isNil, omitBy } from 'lodash-es';
+import { combineLatest, map, switchMap } from 'rxjs';
+import { AbstractType as YAbstractType } from 'yjs';
+
+import { Store } from '../../../framework';
+import {
+  yjsObserveByPath,
+  yjsObserveDeep,
+} from '../../../utils/yjs-observable';
+import type { WorkspaceDBService } from '../../db';
+import type {
+  DocCustomPropertyInfo,
+  DocProperties,
+} from '../../db/schema/schema';
+import type { WorkspaceService } from '../../workspace';
+
+interface LatencyDocProperties {
+  custom?: Record<string, { value: unknown } | undefined>;
+  system?: Record<string, { value: unknown } | undefined>;
+}
+
+type LatencyDocPropertyInfo = {
+  id?: string;
+  name?: string;
+  type?: string;
+};
+
+type LatencyDocPropertyInfoList = Record<
+  string,
+  LatencyDocPropertyInfo | undefined
+>;
+
+export class DocPropertiesStore extends Store {
+  constructor(
+    private readonly workspaceService: WorkspaceService,
+    private readonly dbService: WorkspaceDBService
+  ) {
+    super();
+  }
+
+  updateDocProperties(id: string, config: Partial<DocProperties>) {
+    return this.dbService.db.docProperties.create({
+      id,
+      ...config,
+    });
+  }
+
+  getDocPropertyInfoList() {
+    const db = this.dbService.db.docCustomPropertyInfo.find({
+      isDeleted: false,
+    });
+    const latency = this.upgradeLatencyDocPropertyInfoList(
+      this.getLatencyDocPropertyInfoList()
+    );
+    const notOverridden = differenceBy(latency, db, i => i.id);
+    return [...db, ...notOverridden];
+  }
+
+  createDocPropertyInfo(config: DocCustomPropertyInfo) {
+    return this.dbService.db.docCustomPropertyInfo.create(config).id;
+  }
+
+  removeDocPropertyInfo(id: string) {
+    this.updateDocPropertyInfo(id, {
+      additionalData: {}, // also remove additional data to reduce size
+      isDeleted: true,
+    });
+  }
+
+  updateDocPropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
+    const needMigration = !this.dbService.db.docCustomPropertyInfo.get(id);
+    if (needMigration) {
+      // if this property is not in db, we need to migration it from latency to db, only type and name is needed
+      const latencyId = id.replace('custom:', '');
+      const latency = this.getLatencyDocPropertyInfo(latencyId);
+      this.dbService.db.docCustomPropertyInfo.create({
+        id,
+        type:
+          latency?.type ??
+          'unknown' /* should never reach here, just for safety, we need handle unknown property type */,
+        name: latency?.name,
+        ...config,
+      });
+      return;
+    } else {
+      this.dbService.db.docCustomPropertyInfo.update(id, config);
+    }
+  }
```

Comment:
may split the migration code for clearity

```suggestion
updateDocPropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
  const needMigration = !this.dbService.db.docCustomPropertyInfo.get(id);
  return needMigration
    ? this.migrateAndCreatePropertyInfo(id, config)
    : this.dbService.db.docCustomPropertyInfo.update(id, config);
}

private migrateAndCreatePropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
  const latencyId = id.replace('custom:', '');
  const latency = this.getLatencyDocPropertyInfo(latencyId);
  return this.dbService.db.docCustomPropertyInfo.create({
    id,
    type: latency?.type ?? 'unknown',
    name: latency?.name,
    ...config,
  });
}
```